### DISCUSSION TITLE
Is it possible to backup retained messages via ctl data export?

### DISCUSSION THREAD

**Discussion by evekto** [emqx/emqx#11938] [Q&A] [answered]
Hello👋,

I would like to perform backups with the command line tool as described here: https://www.emqx.io/docs/en/latest/operations/backup-restore.html

Unfortunaley, retained messages are not part of the backup, even after I configured their storage in the build-in database and "disc" persistence as storage type.

I tried to dive a little into the source code and found that database tables for the backup need to have the 'emqx_db_backup' behavior, for example the [emqx_psk module](https://github.com/emqx/emqx/blob/518b02fc704162714990888274f602b38b7b5b59/apps/emqx_psk/src/emqx_psk.erl#L20).

Is there a reason that this behavior is not implemented for the retainer?

Would it suffice to include the line "-behaviour(emqx_config_backup)." or is there more to do?

Thanks for any input on this topic! :-)

**Comment by savonarola** [comment, marked as answer]
Hello!

> Would it suffice to include the line "-behaviour(emqx_config_backup)." or is there more to do?

There is a bit more. `-behaviour(emqx_config_backup).` implies that some callbacks described in `emqx_config_backup` will also be implemented. Luckily, there is only one [callback](https://github.com/emqx/emqx/blob/518b02fc704162714990888274f602b38b7b5b59/apps/emqx/src/bhvrs/emqx_db_backup.erl) returning the list of tables to backup.

So, a function also should be added and exported to the appropriate module, like
```
-export([
  backup_tables/0 
]).

...

backup_tables() -> [?TAB_INDEX_META, ?TAB_INDEX, ?TAB_MESSAGE].
```

See the other modules implementing this.
 
However, in the backup, we backup only configurational data as well as some "configuration-related" tables to keep the backup/restore process robust.
The message data, on the other hand, may reach considerable volumes.

**Reply by evekto** [reply]
Thank you very much!
Considering the increase in the backup's size, an extra config option to explicitly include retained messages in the backup might make sense.
Thanks for indicating what would be needed to implement that.
